source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -158,55 +183,162 @@ LocalRateLimiterImpl::Result LocalRateLimiterImpl::requestAllowed(
+  for (const auto& match_result : matched_results) {
+    if (!match_result.token_bucket->consume(
             share_factor, match_result.request_descriptor.get().hits_addend_.value_or(1))) {
       // If the request is forbidden by a descriptor, return the result and the descriptor
       // token bucket.
-      return {false, makeOptRef<TokenBucketContext>(match_result.token_bucket.get())};
+      return {false, std::shared_ptr<TokenBucketContext>(match_result.token_bucket)};
     }
+    ENVOY_LOG(trace,
+              "request allowed by descriptor with fill rate: {}, maxToken: {}, remainingToken {}",
+              match_result.token_bucket->fillRate(), match_result.token_bucket->maxTokens(),
+              match_result.token_bucket->remainingTokens());
   }
 
   // See if the request is forbidden by the default token bucket.
   if (matched_results.empty() || always_consume_default_token_bucket_) {
     if (default_token_bucket_ == nullptr) {
       return {true, matched_results.empty()
-                        ? makeOptRefFromPtr<TokenBucketContext>(nullptr)
-                        : makeOptRef<TokenBucketContext>(matched_results[0].token_bucket.get())};
+                        ? std::shared_ptr<TokenBucketContext>(nullptr)
+                        : std::shared_ptr<TokenBucketContext>(matched_results[0].token_bucket)};
     }
     ASSERT(default_token_bucket_ != nullptr);
 
     if (const bool result = default_token_bucket_->consume(share_factor); !result) {
       // If the request is forbidden by the default token bucket, return the result and the
       // default token bucket.
-      return {false, makeOptRefFromPtr<TokenBucketContext>(default_token_bucket_.get())};
+      return {false, std::shared_ptr<TokenBucketContext>(default_token_bucket_)};
     }
 
     // If the request is allowed then return the result the token bucket. The descriptor
     // token bucket will be selected as priority if it exists.
-    return {true, makeOptRef<TokenBucketContext>(matched_results.empty()
-                                                     ? *default_token_bucket_
-                                                     : matched_results[0].token_bucket.get())};
+    return {true,
+            matched_results.empty() ? default_token_bucket_ : matched_results[0].token_bucket};
   };
 
   ASSERT(!matched_results.empty());
-  return {true, makeOptRef<TokenBucketContext>(matched_results[0].token_bucket.get())};
+  std::shared_ptr<TokenBucketContext> bucket_context =
+      std::shared_ptr<TokenBucketContext>(matched_results[0].token_bucket);
+  return {true, bucket_context};
+}
+
+// Compare the request descriptor entries with the user descriptor entries. If all non-empty user
+// descriptor values match the request descriptor values, return true
+bool DynamicDescriptorMap::matchDescriptorEntries(
+    const std::vector<RateLimit::DescriptorEntry>& request_entries,
+    const std::vector<RateLimit::DescriptorEntry>& user_entries) {
```

Comment:
The user_entries is a little unclear. It would be better to call it `config_entries` or something.